azure/services/privatedns/privatedns.go

```diff
@@ -20,202 +20,143 @@ import (
-			}
-			if !isVnetLinkManaged {
-				log.V(2).Info("Skipping vnet link deletion for unmanaged vnet link", "vnet link", linkSpec.LinkName, "private dns zone", zoneSpec.ZoneName)
-				continue
-			}
-			log.V(2).Info("removing virtual network link", "virtual network", linkSpec.VNetName, "private dns zone", zoneSpec.ZoneName)
-			err = s.client.DeleteLink(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, linkSpec.LinkName)
-			if err != nil && !azure.ResourceNotFound(err) {
-				return errors.Wrapf(err, "failed to delete virtual network link %s with zone %s in resource group %s", linkSpec.VNetName, zoneSpec.ZoneName, s.Scope.ResourceGroup())
-			}
-		}
-		// Skip the deletion of private DNS zone which is not managed by capz.
-		isManaged, err := s.isPrivateDNSManaged(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName)
-		if err != nil && !azure.ResourceNotFound(err) {
-			return errors.Wrapf(err, "could not get private DNS zone state of %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
-		}
-		if !isManaged {
-			log.V(1).Info("Skipping private DNS zone deletion for unmanaged private DNS zone", "private DNS", zoneSpec.ZoneName)
-			return nil
-		}
-		// Delete the private DNS zone, which also deletes all records.
-		log.V(2).Info("deleting private dns zone", "private dns zone", zoneSpec.ZoneName)
-		err = s.client.DeleteZone(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName)
-		if err != nil && azure.ResourceNotFound(err) {
-			// already deleted
-			return nil
-		}
-		if err != nil && !azure.ResourceNotFound(err) {
-			return errors.Wrapf(err, "failed to delete private dns zone %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
-		}
-		log.V(2).Info("successfully deleted private dns zone", "private dns zone", zoneSpec.ZoneName)
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	zoneSpec, links, _ := s.Scope.PrivateDNSSpec()
+	if zoneSpec == nil {
+		return nil
 	}
-	return nil
+
+	managed, err := s.deleteLinks(ctx, links)
+	if managed {
+		s.Scope.UpdateDeleteStatus(infrav1.PrivateDNSLinkReadyCondition, serviceName, err)
+	}
+	if err != nil {
+		return err
+	}
+
+	managed, err = s.deleteZone(ctx, zoneSpec)
+	if managed {
+		s.Scope.UpdateDeleteStatus(infrav1.PrivateDNSZoneReadyCondition, serviceName, err)
+		s.Scope.UpdateDeleteStatus(infrav1.PrivateDNSRecordReadyCondition, serviceName, err)
+	}
+
+	return err
 }
 
 // isPrivateDNSManaged returns true if the private DNS has an owned tag with the cluster name as value,
 // meaning that the DNS lifecycle is managed.
-func (s *Service) isPrivateDNSManaged(ctx context.Context, resourceGroup, zoneName string) (bool, error) {
-	zone, err := s.client.GetZone(ctx, resourceGroup, zoneName)
+func (s *Service) isPrivateDNSManaged(ctx context.Context, spec azure.ResourceSpecGetter) (bool, error) {
```

Comment:
I see what you're saying about the redundant Get() calls too, it looks like we have this pattern on several services like VNets, where we need to fetch the existing resource to see if it's managed or not. I'm open to amending the interface, did you have any ideas? Since the tags are all the same, I was thinking we could have some an interface function where it could cache the result for the next Get() call for Create(). Wdyt @shysank @CecileRobertMichon?

Reply:
I'm inclined towards merging this one as is and updating the interface in a followup. Since it affects only private clusters, this shouldn't cause huge performance degradation imo.

Reply:
yes +1 to caching as a separate issue, the scope is big enough here. #2146 (the second commit for IsManaged) should also help with that. I had tried to incorporate caching as part of https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684/files#diff-b74c6acb4629abe58f8286da9939563cc4404e09fe9966eb61dbde9b87274189R110 but ended up reverting it to keep the changes small.